torch/distributed/pipelining/stage.py

```diff
@@ -649,6 +649,11 @@ def run_post_backward(fsdp_module: FSDPModule) -> None:
                         if state._fsdp_param_group:
                             state._fsdp_param_group.post_backward()
 
+                    # it would be much better if pipelining backward invoked .backward so autograd hooks
+                    # worked and modules like DDP/FSDP behaved as expected.  Working around this for the time being,
+                    # we need to call this too to ensure FSDP syncs its grad reduction ops back to the default stream.
+                    fsdp_state._root_post_backward_final_callback()
```

Comment:
What if `perform_backward` calls `stage_backward`? Do we get double hook in such a case?

Reply:
it appears we do. However, according to @awgu it is idempotent so the second call should not be a problem.

Still, this all points to the design being wrong.  We should find a way to make PP invoke .backward() consistently so we don't have to deal with hooks/accumulations manually.

cc @H-Huang

Reply:
sorry, i was confused, but there is no double-hook issue because we _always_ disable the FSDP hooks before calling into perform_backward.  So this PR should be correct, and it should fix correctness for both Zero-bubble and non-zero-bubble schedules since we _always_ manually call FSDP hooks.
<img width="1236" alt="Screenshot 2025-01-10 at 3 15 55 PM" src="https://github.com/user-attachments/assets/18f1171f-1d73-42f6-b783-5cfb54399d28" />